### PR TITLE
Add sleep callbacks to Asyncify, allowing SDL Audio to automatically update

### LINKED ISSUE #9823: Asyncify behaving differently then emterpreter
Hi. This is not strictly bug, I just found use case where asyncify works different then emterpreter (works badly).
Current documentation says:
```
The Emterpreter has “yielding” as a concept, but it isn’t needed in Asyncify. 
You can replace emscripten_sleep_with_yield() calls with emscripten_sleep().
```

It's almost true, but for sdl sound library is not correct.

For fastcomp `emscripten_sleep_with_yeld` is also pushing new audio data:
```
var yieldCallback = function() {
    if (SDL.audio && SDL.audio.queueNewAudioData) SDL.audio.queueNewAudioData()
};
SDL.audio.yieldCallback = yieldCallback;
EmterpreterAsync.yieldCallbacks.push(yieldCallback);
```

But for asyncify `emsripten_sleep` doesn't do that.

Because of that this conceptual program will work fine for emterpeter but not for asyncify:

```
SDL_AudioSpec as;
as.callback = audio_callback;

void audio_callback(void *unused, Uint8 *stream, int len) {
   // push audio stream data to stream variable
}

while (true) {
   // calculate audio stream data
   emscripten_sleep_with_yeld(1); // emscripten_sleep(1); - for asyncify
}
```

In asyncify sound will be distorted. Anyway this can be esily fixed for asyncify too:

```
while (true) {
    // calculate audio stream data
    EM_ASM(({
        SDL.audio.queueNewAudioData();
    }));
    emscripten_sleep(1);
}
```

So, maybe to resolve this issue is better to add this snippet in documentation.
I found this bug in dosbox, because sound processing is doing in this way.
[Arkanoid](https://js-dos.com/6.22/examples/?arkanoid) will play sounds normally in emterpreter mode and distorted in asyncify.

Also I have minimal test that demonstrate problem:
[sdl_audio_asyncify.zip](https://github.com/emscripten-core/emscripten/files/3835494/sdl_audio_asyncify.zip)

### LINKED ISSUE #9827: ENV and IDBFS are undefined in 1.38.47-upstream & above
Reporting an issue we are seeing as in the Qt bug report:
https://bugreports.qt.io/browse/QTBUG-79971

ENV and IDBFS javascript use in upstream (1.38.47, 1.39.2) do not work as they did, and result in the message:

ReferenceError: ENV is not defined qtloader.js:411:17

### PR DESCRIPTION
**Pull request by kripken**
This fixes a regression compared to fastcomp/emterpretify.

Enables `browser.test_sdl_audio_beep_sleep` on upstream.

Fixes #9823

### CODE REVIEW

**Comment by VirtualTim** [comment]
I'm pretty sure that commit message is wrong. This should fix 9823, not 9827.

**Comment by kripken** [comment]
Correct, thanks @VirtualTim !

File: src/library_async.js [hunk 1]
```diff
@@ -723,6 +724,9 @@ mergeInto(LibraryManager.library, {
             asyncFinalizers.forEach(function(func) {
               func(asyncWasmReturnValue);
             });
+            Asyncify.sleepCallbacks.forEach(function(func) {
```
Comment (sbc100) [inline src/library_async.js]: Why qualify this with `Asyncify.` but not `asyncFinalizers` three lines above?

File: tests/sdl_audio_beep_sleep.cpp [hunk 2]
```diff
@@ -208,7 +212,7 @@ void nextTest(void *unused = 0) {
 
 void update() {
   SDL_LockAudio();
-  int size = beep->beeps.size();
+  int size = beep ? beep->beeps.size() : 0;
```
Comment (sbc100) [inline tests/sdl_audio_beep_sleep.cpp]: unrelated?

File: src/library_async.js (hunk 1 above)
Comment (caiiiycuk) [inline src/library_async.js]: This never works in dosbox case,  cause condition on line 709 **never** became true.
```
          if (!Asyncify.currData) {
```
image
if I move `Asyncify.sleepCallbacks.forEach` before this everything is fixed. Actually I don't know what meaning of this currData.

Comment (caiiiycuk) [reply src/library_async.js]: Even this test code didn't work:
```
#include <SDL/SDL.h>
#include <stdio.h>
#include <sys/time.h>
#include <algorithm>
using namespace std;

#ifdef EMSCRIPTEN
#include <emscripten.h>
#endif

long int reportedSec = 0;
struct timeval startTime;
int available = 0;

void func_callback(void *unused, Uint8 *stream, int len) {

    for (int i=0;i<std::min(len, available);i++) {
        stream[i] = i;
    }

    for (int i = available; i < len; ++i) {
        stream[i] = 0;
    }
    available = 0;
}

void loop() {
    struct timeval nowTime;
    gettimeofday(&nowTime, NULL);

    if (nowTime.tv_sec != reportedSec) {
        reportedSec = nowTime.tv_sec;
        printf("playing %ld sec...\n", reportedSec - startTime.tv_sec);
    }

    for (int i = 0; i < 15; ++i) {
        available = 1024;
        #ifdef EMSCRIPTEN
            //EM_ASM(({
            //    SDL.audio.queueNewAudioData();
            //}));
            emscripten_sleep(10);
        #endif
    }
}

int main(int argc, char *argv[])
{
    gettimeofday(&startTime, NULL);
    if( SDL_Init(SDL_INIT_TIMER | SDL_INIT_AUDIO ) <0 ) {
        SDL_WM_SetCaption( "Something busy..", NULL );
        return 1;
    }

    SDL_AudioSpec as;

    as.freq = 22000;
    as.format = AUDIO_U8;
    as.samples = 1024;
    as.callback=func_callback;
    as.userdata=NULL;
    as.channels = 1;
    
    if ( SDL_OpenAudio(&as, NULL) < 0 ) {
        SDL_WM_SetCaption( "Unable to open audio..", NULL );
        return 1;
    }

    SDL_Surface *screen = SDL_SetVideoMode(200,200, 16, SDL_SWSURFACE);
    
    SDL_PauseAudio(0);

#ifdef EMSCRIPTEN
    emscripten_set_main_loop(loop, 0, true);
#endif

    return 0;
}
```

Comment (kripken) [reply src/library_async.js]: `asyncFinalizers` is captured to a local variable because we clear it out. Those finalizer calls are executed once. The `sleepCallbacks` on the other hand happen constantly, every time.

Comment (kripken) [reply src/library_async.js]: Thanks @caiiiycuk ! Yeah, that code was not in the right place, I realize now.

I pushed an update, please tell me if that fixes it.

I tested your testcase here, and it prints "playing 0", 1, 2, 3 etc., and buzzes, does that mean it works?

Comment (caiiiycuk) [reply src/library_async.js]: @kripken Yeah this fix works for dosobox and testcase. Thanks!

--
Testcase works if it constantly buzzes without hangs.
